mmr/runner.rs

```diff
@@ -72,37 +58,40 @@ impl Runner {
         }
     }
 
+    /// Get the cache count
+    // fn cache_count(&self) -> Result<i64, Error> {
+    //     let store = Store::new(&self.path);
+    //     let res = eth_header_with_proof_caches
+    //         .filter(root.is_not_null())
+    //         .select(count(root))
+    //         .first::<i64>(&store.conn);
+    //     if let Err(e) = res {
+    //         Err(Error::Diesel(e))
+    //     } else {
+    //         Ok(res?)
+    //     }
+    // }
+
     /// Push new header hash into storage
     fn push(&mut self, pnumber: i64) -> Result<(), Error> {
         let store = Store::new(&self.path);
-        let count = self.mmr_count().unwrap_or(0);
-
         // Get Hash
         let conn = store.conn();
         let cache = eth_header_with_proof_caches
             .filter(number.eq(pnumber))
             .first::<Cache>(&store.conn)?;
+        println!("{:?}", cache);
 
-        let mut mmr = MMR::<_, MergeHash, _>::new(count as u64, store);
-        let rpos = mmr.push(H256::from(&cache.hash[2..]))?;
-
-        // gen mmr proof
-        let proof = mmr
-            .gen_proof(vec![rpos])?
-            .proof_items()
-            .iter()
-            .map(|h| h.hex())
-            .collect::<Vec<String>>()
-            .join(",");
+        let mut mmr =
+            MMR::<_, MergeHash, _>::new(cmmr::leaf_index_to_mmr_size(pnumber as u64) - 1, store);
```

Comment:
This part might have ambiguities.

If use the `mmr_size` to generate MMR before we push the `new leaf` into the store, an Error `MMR(InconsistentStore) ` will be fired, So if we shoud:

+ use `leaf_index_to_mmr_size(pnumber as u64 - 1)` when we push a new leaf into store
+ use `leaf_index_to_mmr_size(pnumber as u64)` while generating proof

Reply:
Shoud not be `mmr_size - 1`, shoud be `last_mmr_size + 1`

https://github.com/nervosnetwork/merkle-mountain-range/blob/ea232474496bdf072d7de33297fd54f1ef8df599/src/mmr.rs#L70

Reply:
should be last_mmr_size, self.mmr_count()

Reply:
https://github.com/darwinia-network/shadow/issues/34

Reply:
leaf_index_to_mmr_size is used when gen_proof, to retrieve a sub MMR for generating proof for (member_index, last_index), here before append, should use node counts in the store.

Reply:
> #34 

#34 is one of the BUGs, and there must be more fishy BUGs...The common tests work fine(using the database this repo implemented), but the production test doesn't.